### PR TITLE
Added NSE monitoring to NSM

### PR DESCRIPTION
**Pull request by Bolodya1997**
## Description

Added NSE monitoring to NSM. If connection to NSE dies, NSE is getting deleted from the registry.
If connection to NSE is not getting alive in 10 seconds, NSE is getting deleted from the registry.

## Motivation and Context

NSMD should periodically check NSE without connections are alive and update registry since field. #836

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [x] Added integration testing to cover
- [x] Tested locally
- [ ] Have not tested

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: controlplane/pkg/nsmd/nsmd.go [hunk 1]
```diff
@@ -280,6 +264,46 @@ func (nsm *nsmServer) restoreClients(registeredEndpoints *registry.NetworkServic
...
+func (nsm *nsmServer) deleteEndpoint(name string) error {
+	client, err := nsm.serviceRegistry.NseRegistryClient()
+	if err != nil {
+		return err
+	}
+
+	return nsm.deleteEndpointWithClient(name, client)
+}
+
+func (nsm *nsmServer) deleteEndpointWithClient(name string, client registry.NetworkServiceRegistryClient) error {
```
Comment (lobkovilya) [inline controlplane/pkg/nsmd/nsmd.go]: Maybe you know what happens with endpoint's workspace? Is it deleted by DevicePlugin, or we should delete workspace here?

File: controlplane/pkg/nsmd/nsmd_crossconnect_client_test.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		endpointStateManager)
+	model.AddListener(client)
+
+	model.AddEndpoint(createEndpoint("ws1", "nse1"))
+	model.AddEndpoint(createEndpoint("ws2", "nse2"))
+	model.AddEndpoint(createEndpoint("ws3", "nse3"))
+
+	endpointStateManager.KillEndpoint("ws1", "nse1")
+
+	_ = model.DeleteEndpoint("nse2")
```
Comment (lobkovilya) [inline controlplane/pkg/nsmd/nsmd_crossconnect_client_test.go]: Same, unnecessary assigning

File: controlplane/pkg/nsmd/nsmd_crossconnect_client_test.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	endpointStateManager.KillEndpoint("ws2", "nse2")
+
+	time.Sleep(1 * time.Second)
```
Comment (lobkovilya) [inline controlplane/pkg/nsmd/nsmd_crossconnect_client_test.go]: Can we avoid using sleep?

File: test/kube_testing/k8s.go [hunk 4]
```diff
@@ -603,3 +604,11 @@ func (o *K8s) CleanupConfigMaps() {
 		_ = o.clientset.CoreV1().ConfigMaps("default").Delete(cm.Name, &metaV1.DeleteOptions{})
```
Comment (lobkovilya) [inline test/kube_testing/k8s.go]: Same, unnecessary assigning

File: test/integration/basic_delete_dirty_nse_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+	Expect(err).To(BeNil())
+
+	k8s.PrepareDefault()
+	nodes := nsmd_test_utils.SetupNodesConfig(k8s, 1, defaultTimeout, []*pods.NSMgrPodConfig{})
+
+	nsePod := nsmd_test_utils.DeployDirtyNSE(k8s, nodes[0].Node, "dirty", defaultTimeout)
+	k8s.DeletePods(nsePod)
+
+	time.Sleep(5 * time.Second)
```
Comment (lobkovilya) [inline test/integration/basic_delete_dirty_nse_test.go]: And here?

File: controlplane/pkg/nsmd/nsmd_crossconnect_client.go [hunk 6]
```diff
@@ -134,6 +156,23 @@ func (client *NsmMonitorCrossConnectClient) ClientConnectionDeleted(clientConnec
...
 }
 
+func (client *NsmMonitorCrossConnectClient) endpointConnectionMonitor(endpoint *model.Endpoint, ctx context.Context) {
+	logrus.Infof("NSM-EndpointMonitor(%v): Started", endpoint.EndpointName())
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Infof("NSM-EndpointMonitor(%v): Context timeout exceeded...", endpoint.EndpointName())
+			return
+		case <-time.Tick(3):
```
Comment (lobkovilya) [inline controlplane/pkg/nsmd/nsmd_crossconnect_client.go]: Probably `3 * time.Second`?

Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd_crossconnect_client.go]: Fixed

File: test/integration/basic_delete_dirty_nse_test.go (hunk 5 above)
Comment (Bolodya1997) [reply test/integration/basic_delete_dirty_nse_test.go]: Fixed : [place](https://github.com/Bolodya1997/networkservicemesh/blob/nsm-monitor-nse/test/integration/basic_delete_dirty_nse_test.go#L73)

File: test/kube_testing/k8s.go (hunk 4 above)
Comment (Bolodya1997) [reply test/kube_testing/k8s.go]: The line is out of the PR changes

File: controlplane/pkg/nsmd/nsmd_crossconnect_client_test.go (hunk 3 above)
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd_crossconnect_client_test.go]: Fixed : [place](https://github.com/Bolodya1997/networkservicemesh/blob/nsm-monitor-nse/controlplane/pkg/nsmd/nsmd_crossconnect_client_test.go#L47)

File: controlplane/pkg/nsmd/nsmd_crossconnect_client_test.go (hunk 2 above)
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd_crossconnect_client_test.go]: It is for suppressing GoLand 'Unhandled error' check

File: controlplane/pkg/nsmd/nsmd.go (hunk 1 above)
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd.go]: As discussed offline, workspace should be deleted by DevicePlugin.

File: controlplane/pkg/nsmd/nsmd_crossconnect_client.go [hunk 7]
```diff
@@ -134,6 +156,23 @@ func (client *NsmMonitorCrossConnectClient) ClientConnectionDeleted(clientConnec
...
+			if alive, err := client.endpointStateManager.isEndpointAlive(endpoint.Workspace, endpoint.EndpointName()); err != nil {
+				client.EndpointDeleted(endpoint)
+			} else if !alive {
+				_ = client.endpointStateManager.deleteEndpoint(endpoint.EndpointName())
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd_crossconnect_client.go]: Same, suppressing check.

[the remark this answers is not in the document]
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd_crossconnect_client.go]: Fixed : [place](https://github.com/Bolodya1997/networkservicemesh/blob/nsm-monitor-nse/controlplane/pkg/nsmd/nsmd_crossconnect_client.go#L199)

File: controlplane/pkg/nsmd/nsmd.go [hunk 8]
```diff
@@ -280,6 +264,46 @@ func (nsm *nsmServer) restoreClients(registeredEndpoints *registry.NetworkServic
...
+	}
+
+	ctx, cancelCtx := context.WithTimeout(context.Background(), NSEAliveTimeout)
+	defer cancelCtx()
+
+	logrus.Infof("Checking NSE %s is alive at %v...", name, ws.NsmClientSocket())
+	if nseConn, err := tools.SocketOperationCheckContext(ctx, tools.SocketPath(ws.NsmClientSocket())); err != nil {
+		return false, nil
+	} else {
+		_ = nseConn.Close()
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd.go]: Same, supressing check.

File: controlplane/pkg/nsmd/endpoint_state_manager_stub.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+import model_ "github.com/networkservicemesh/networkservicemesh/controlplane/pkg/model"
+
+type EndpointStateManagerStub struct {
+	endpoints map[string]*model_.Endpoint
+	model     model_.Model
+	model_.ModelListenerImpl
+}
+
+func NewEndpointStateManagerStub(model model_.Model) *EndpointStateManagerStub {
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/endpoint_state_manager_stub.go]: Fixed : [place](https://github.com/Bolodya1997/networkservicemesh/blob/nsm-monitor-nse/controlplane/pkg/nsmd/endpoint_state_manager_stub.go#L11)

File: test/kube_testing/pods/dirty_nse.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+			Containers: []v1.Container{
+				{
+					Name:            "dirty-nse",
+					Image:           "networkservicemesh/nse:latest",
+					ImagePullPolicy: v1.PullIfNotPresent,
+					Resources: v1.ResourceRequirements{
+						Limits: v1.ResourceList{
+							"networkservicemesh.io/socket": resource.NewQuantity(1, resource.DecimalSI).DeepCopy(),
+						},
+						Requests: nil,
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply test/kube_testing/pods/dirty_nse.go]: Fixed in all pod description files : c89a02c

File: controlplane/pkg/nsmd/nsmd.go [hunk 11]
```diff
@@ -280,6 +264,46 @@ func (nsm *nsmServer) restoreClients(registeredEndpoints *registry.NetworkServic
...
 }
 
+func (nsm *nsmServer) isEndpointAlive(workspace string, name string) (bool, error) {
+	ws, ok := nsm.workspaces[workspace]
+	if !ok {
+		return false, fmt.Errorf("workspace %v doesn't exist", workspace)
+	}
+
+	ctx, cancelCtx := context.WithTimeout(context.Background(), NSEAliveTimeout)
+	defer cancelCtx()
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd.go]: Seems like there is no need for this, removed : c89a02c

File: .mk/docker.mk [hunk 12]
```diff
@@ -14,7 +14,7 @@
 
 BUILD_CONTAINERS=nsmd nsmdp nsmd-k8s vppagent-dataplane vppagent-dataplane-dev
 BUILD_CONTAINERS+=devenv crossconnect-monitor
-BUILD_CONTAINERS+=nsc icmp-responder-nse
+BUILD_CONTAINERS+=nsc nse
```
Comment (nickolaev) [inline .mk/docker.mk]: Why the rename? Can you elaborate in the PR description?

Comment (nickolaev) [reply .mk/docker.mk]: Also there are quite a few files changed with this rename, maybe it is a better idea to split these to another PR.

File: examples/cmd/nse/nse_run.sh [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+trap sigint INT
+trap sigterm TERM
+trap sigquit QUIT
+
+echo Starting NSE: "${NSE_IMAGE}"
+"/bin/${NSE_IMAGE}" &
+
+NSE=$!
+wait "${NSE}"
+wait "${NSE}"
```
Comment (nickolaev) [inline examples/cmd/nse/nse_run.sh]: Why 2 times `wait` here?

File: examples/cmd/nse/dirty-nse/nse.go [hunk 14]
```diff
@@ -0,0 +1,27 @@
+package main
+
+import (
+	"time"
+
+	"github.com/networkservicemesh/networkservicemesh/sdk/endpoint"
+	"github.com/sirupsen/logrus"
+)
+
+func main() {
```
Comment (nickolaev) [inline examples/cmd/nse/dirty-nse/nse.go]: I am not sure why is this needed?
Also, please add signal handling.

File: examples/cmd/nse/nse_run.sh (hunk 13 above)
Comment (Bolodya1997) [reply examples/cmd/nse/nse_run.sh]: When the container stops, the shell process receives the signal, breaks `wait` execution, sends the received signal to the NSE process with the `kill` and then it should wait for the NSE process to complete all shutdown hooks.
If we omit this second wait, the NSE process will also receive the signal but the container would be destroyed before the shutdown hooks got executed.

File: examples/cmd/nse/dirty-nse/nse.go (hunk 14 above)
Comment (Bolodya1997) [reply examples/cmd/nse/dirty-nse/nse.go]: In this context **dirty** means that NSE doesn't perform any cleanup routines. There are a lot of cases when the process can suddenly stop and never continue the execution, so it is to simulate such cases.
So I believe there is no need to add some signal handling in here. The process is broken and so it should be killed by the signal received.

File: .mk/docker.mk (hunk 12 above)
Comment (Bolodya1997) [reply .mk/docker.mk]: Separate PR : #1013

**Comment by haiodo** [comment]
I suppose this on is WIP and will be separated to few PRs. @Bolodya1997 could you please rename it to include WIP:

**Comment by edwarnicke** [comment]
@haiodo +1

File: controlplane/pkg/nsmd/nsmd_crossconnect_client.go [hunk 15]
```diff
@@ -134,6 +162,44 @@ func (client *NsmMonitorCrossConnectClient) ClientConnectionDeleted(clientConnec
...
+		case <-time.Tick(endpointLivenessTimeout):
+			connected = client.endpointStateManager.IsEndpointAlive(endpoint)
+		}
+	}
+
+	if connected {
+		logrus.Infof(endpointLogFormat, endpoint.EndpointName(), "Connected")
+	}
+
+	for connected {
```
Comment (haiodo) [inline controlplane/pkg/nsmd/nsmd_crossconnect_client.go]: So here, you need to check if it alive forever until it had active connection or alive.

File: controlplane/pkg/nsmd/nsmd.go [hunk 16]
```diff
@@ -278,6 +270,55 @@ func (nsm *nsmServer) restoreClients(registeredEndpoints *registry.NetworkServic
 	logrus.Infof("NSMD: Restore of NSE/Clients Complete...")
 }
 
+func (ws *Workspace) isAlive() bool {
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd.go]: Renamed this to `isConnectionAlive` and moved to `workspace.go`, so it is both good as for NSE so for NSC case : 2f1a9ed

File: controlplane/pkg/nsmd/nsmd_crossconnect_client.go [hunk 17]
```diff
@@ -134,6 +162,44 @@ func (client *NsmMonitorCrossConnectClient) ClientConnectionDeleted(clientConnec
 	}
 }
 
+func (client *NsmMonitorCrossConnectClient) endpointConnectionMonitor(endpoint *model.Endpoint, ctx context.Context) {
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd_crossconnect_client.go]: I suppose that on the **monitor** abstraction level there are only 2 states for the endpoint: state to keep and state to remove.
The layer responsive to determine the exact state of the endpoint (has active connection / doesn't have active connection + alive / not alive) can first check for the active connections and only after check if the connection if still alive : 43b0835

File: controlplane/pkg/nsmd/nsmd_crossconnect_client.go (hunk 15 above)
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd_crossconnect_client.go]: It is actually the thing that I am doing here. Probably `IsEndpointAlive()` was not so meaningful name, so I have changed it to `IsEndpointAliveOrHasActiveConnections()` : 43b0835

File: controlplane/pkg/nsmd/nsmd_crossconnect_client.go [hunk 18]
```diff
@@ -36,28 +36,43 @@ import (
 	"google.golang.org/grpc"
 )
 
+const (
+	endpointLogFormat       = "NSM-EndpointMonitor(%v): %v"
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply controlplane/pkg/nsmd/nsmd_crossconnect_client.go]: Filed an issue for this : #1053

**Comment by Bolodya1997** [comment]
This PR is based on #1146 and #1148 and so comes after.

File: examples/cmd/nse/icmp-responder-nse/nse.go [hunk 19]
```diff
@@ -72,6 +79,11 @@ func main() {
 		defer func() { _ = nsmEndpoint.Delete() }()
 	}
 
+	if *update {
+		<-time.After(10 * time.Second)
```
Comment (haiodo) [inline examples/cmd/nse/icmp-responder-nse/nse.go]: Why we wait here 10 seconds?

Comment (Bolodya1997) [reply examples/cmd/nse/icmp-responder-nse/nse.go]: To be sure that client has connected to NSM and has started monitoring connection update events.

Comment (haiodo) [reply examples/cmd/nse/icmp-responder-nse/nse.go]: We need to send immediately and handle on init state properly if state are changed we have updated on NSC.

Comment (Bolodya1997) [reply examples/cmd/nse/icmp-responder-nse/nse.go]: Done : 43e3fdb

**Comment by Bolodya1997** [comment]
Depends on #1189.
